server/src/main/java/org/opensearch/cluster/metadata/Metadata.java

```diff
@@ -1139,8 +1151,17 @@ public Builder(Metadata metadata) {
             this.hashesOfConsistentSettings = metadata.hashesOfConsistentSettings;
             this.version = metadata.version;
             this.indices = new HashMap<>(metadata.indices);
+            this.indicesPreviousState = new HashMap<>(metadata.indices); // required for comparing with updated indices
             this.templates = new HashMap<>(metadata.templates);
             this.customs = new HashMap<>(metadata.customs);
+            this.customsPreviousState = new HashMap<>(metadata.customs);
+            this.indicesLookup = new TreeMap<>(metadata.indicesLookup);
+            this.allIndices = Arrays.copyOf(metadata.allIndices, metadata.allIndices.length);
+            this.visibleIndices = Arrays.copyOf(metadata.visibleIndices, metadata.visibleIndices.length);
+            this.allOpenIndices = Arrays.copyOf(metadata.allOpenIndices, metadata.allOpenIndices.length);
+            this.visibleOpenIndices = Arrays.copyOf(metadata.visibleOpenIndices, metadata.visibleOpenIndices.length);
+            this.allClosedIndices = Arrays.copyOf(metadata.allClosedIndices, metadata.allClosedIndices.length);
+            this.visibleClosedIndices = Arrays.copyOf(metadata.visibleClosedIndices, metadata.visibleClosedIndices.length);
```

Comment:
1. **There is no change in memory consumption of Metadata object since no new variables are introduced in Metadata class**. These are only introduced in Builder class and memory should get free once Metadata is created and Builder object is sent for GC.

2. In Builder class also, these were initially created inside `build()` invocation. These will not be created again when there are no changes in `indices` and `customs`.